### ISSUE TITLE
[Bug] XmlParsingError   : An internal error occurred while Generate Code [sf#189]

### ISSUE THREAD

**Issue by DietmarSchwertberger** [DietmarSchwertberger/wxGlade#264]
**Reported by frank6363 on 2016-06-14 08:51 UTC**
While trying to generate Perl or Python or C++ code, I get this message on 2 different PC:

~~~
ERROR   : An internal error occurred while Generate Code
An unexpected error occurred!

Date and time:       2016-06-13T16:39:00.894000
Python version:      2.7.11
wxPython version:    3.0.2.0
wxWidgets platform:  __WXMSW__
wxGlade version:     0.7.2
Language code:       French_Belgium
Encoding:            1252
Filesystem encoding: mbcs

Exception type:      <class 'xml_parse.XmlParsingError'>
Exception details:   [Errno 13] Permission denied: u'wxglade_out.pl' (line: 5, column: 0)
Application stack trace:
Stack frame at level 0
======================
  File "C:\Program Files (x86)\wxGlade\codegen\__init__.py", line 2300
  Function "save_file(self=<perl_codegen.PerlCodeWriter object at 0x04098EB0>, filename=u'wxglade_out.pl', content='use Wx 0.15 qw[:allclasses];\nuse strict;\n\n# begin wxGlade: dependencies\n# end wxGlade\n\n# begin wxGlade: extracode\n# end wxGlade\n\n', mainfile=True, content_only=False)"
  Source code context:
              try:
                  common.save_file(filename, tmp, 'codegen')
              except IOError, e:
  ->              raise XmlParsingError(str(e))
              except:
                  self._logger.exception(_('Internal Error'))
              if mainfile and sys.platform in ['linux2', 'darwin']:
  Local variables:
  -> tmp (<type 'str'>): '#!/usr/bin/perl -w -- \n#\n# generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016\n#\n# To get wxPerl visit http://www.wxperl.it\n#\n\nuse Wx 0.15 qw[:allclasses];\nuse strict;\n\n# begin wxGlade: dependencies\n# end wxGlade\n\n# begin wxGlade: extracode\n# end wxGlade\n\n'
  -> mainfile (<type 'bool'>): True
  -> e (<type 'exceptions.IOError'>): IOError(13, 'Permission denied')
  -> self (<class 'perl_codegen.PerlCodeWriter'>): <perl_codegen.PerlCodeWriter object at 0x04098EB0>
  -> filename (<type 'unicode'>): u'wxglade_out.pl'
  -> content (<type 'str'>): 'use Wx 0.15 qw[:allclasses];\nuse strict;\n\n# begin wxGlade: dependencies\n# end wxGlade\n\n# begin wxGlade: extracode\n# end wxGlade\n\n'
  -> content_only (<type 'bool'>): False
  -> dirname (<type 'unicode'>): u''

Stack frame at level 1
======================
  File "C:\Program Files (x86)\wxGlade\codegen\__init__.py", line 1079
  Function "finalize(self=<perl_codegen.PerlCodeWriter object at 0x04098EB0>)"
  Source code context:
                  content = content.replace(tag, code)
      
                  self.output_file.close()
  ->              self.save_file(self.output_file_name, content, self._app_added)
                  del self.output_file
      
          def add_app(self, app_attrs, top_win_class):
  Local variables:
  -> em (<type 'str'>): ''
  -> code (<type 'str'>): '# begin wxGlade: extracode\n# end wxGlade\n'
  -> dep_list (<type 'list'>): []
  -> content (<type 'str'>): 'use Wx 0.15 qw[:allclasses];\nuse strict;\n\n# begin wxGlade: dependencies\n# end wxGlade\n\n# begin wxGlade: extracode\n# end wxGlade\n\n'
  -> tag (<type 'str'>): '<1465828740858345275wxGlade replace extracode>'
  -> self (<class 'perl_codegen.PerlCodeWriter'>): <perl_codegen.PerlCodeWriter object at 0x04098EB0>

Stack frame at level 2
======================
  File "C:\Program Files (x86)\wxGlade\xml_parse.py", line 836
  Function "endElement(self=<xml_parse.CodeWriter instance at 0x0436D760>, name=u'application')"
  Source code context:
                  if self.app_attrs:
                      self.code_writer.add_app(self.app_attrs, self.top_window)
                  # call the finalization function of the code writer
  ->              self.code_writer.finalize()
                  return
              if name == 'object':
                  # remove last object from Stack
  Local variables:
  -> self (<type 'instance'>): <xml_parse.CodeWriter instance at 0x0436D760>
  -> name (<type 'unicode'>): u'application'

Stack frame at level 3
======================
  File "C:\Python27\lib\xml\sax\expatreader.py", line 320
  Function "end_element(self=<xml.sax.expatreader.ExpatParser instance at 0x0436D878>, name=u'application')"
  Source code context:
              self._cont_handler.startElement(name, AttributesImpl(attrs))
      
          def end_element(self, name):
  ->          self._cont_handler.endElement(name)
      
          def start_element_ns(self, name, attrs):
              pair = name.split()
  Local variables:
  -> self (<type 'instance'>): <xml.sax.expatreader.ExpatParser instance at 0x0436D878>
  -> name (<type 'unicode'>): u'application'

Stack frame at level 4
======================
  File "C:\Python27\lib\xml\sax\expatreader.py", line 213
  Function "feed(self=<xml.sax.expatreader.ExpatParser instance at 0x0436D878>, data='<?xml version="1.0"?>\n<!-- generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016 -->\n\n<application class="" encoding="CP1252" for_version="3.0" header_extension=".h" indent_amount="4" indent_symbol="space" is_template="0" language="perl" name="" option="0" overwrite="1" path="wxglade_out.pl" source_extension=".cpp" top_window="" use_gettext="1" use_new_namespace="1">\n</application>\n', isFinal=0)"
  Source code context:
                  # If it is set to true, expat will check validity of the entire
                  # document. When feeding chunks, they are not normally final -
                  # except when invoked from close.
  ->              self._parser.Parse(data, isFinal)
              except expat.error, e:
                  exc = SAXParseException(expat.ErrorString(e.code), e, self)
                  # FIXME: when to invoke error()?
  Local variables:
  -> self (<type 'instance'>): <xml.sax.expatreader.ExpatParser instance at 0x0436D878>
  -> data (<type 'str'>): '<?xml version="1.0"?>\n<!-- generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016 -->\n\n<application class="" encoding="CP1252" for_version="3.0" header_extension=".h" indent_amount="4" indent_symbol="space" is_template="0" language="perl" name="" option="0" overwrite="1" path="wxglade_out.pl" source_extension=".cpp" top_window="" use_gettext="1" use_new_namespace="1">\n</application>\n'
  -> isFinal (<type 'int'>): 0

Stack frame at level 5
======================
  File "C:\Python27\lib\xml\sax\xmlreader.py", line 123
  Function "parse(self=<xml.sax.expatreader.ExpatParser instance at 0x0436D878>, source=<xml.sax.xmlreader.InputSource instance at 0x04382990>)"
  Source code context:
              file = source.getByteStream()
              buffer = file.read(self._bufsize)
              while buffer != "":
  ->              self.feed(buffer)
                  buffer = file.read(self._bufsize)
              self.close()
      
  Local variables:
  -> buffer (<type 'str'>): '<?xml version="1.0"?>\n<!-- generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016 -->\n\n<application class="" encoding="CP1252" for_version="3.0" header_extension=".h" indent_amount="4" indent_symbol="space" is_template="0" language="perl" name="" option="0" overwrite="1" path="wxglade_out.pl" source_extension=".cpp" top_window="" use_gettext="1" use_new_namespace="1">\n</application>\n'
  -> source (<type 'instance'>): <xml.sax.xmlreader.InputSource instance at 0x04382990>
  -> self (<type 'instance'>): <xml.sax.expatreader.ExpatParser instance at 0x0436D878>
  -> file (<type 'instance'>): <StringIO.StringIO instance at 0x043828A0>
  -> saxutils (<type 'module'>): <module 'xml.sax.saxutils' from 'C:\Python27\lib\xml\sax\saxutils.pyc'>

Stack frame at level 6
======================
  File "C:\Python27\lib\xml\sax\expatreader.py", line 110
  Function "parse(self=<xml.sax.expatreader.ExpatParser instance at 0x0436D878>, source=<xml.sax.xmlreader.InputSource instance at 0x04382990>)"
  Source code context:
              self._source = source
              self.reset()
              self._cont_handler.setDocumentLocator(ExpatLocator(self))
  ->          xmlreader.IncrementalParser.parse(self, source)
      
          def prepareParser(self, source):
              if source.getSystemId() is not None:
  Local variables:
  -> source (<type 'instance'>): <xml.sax.xmlreader.InputSource instance at 0x04382990>
  -> self (<type 'instance'>): <xml.sax.expatreader.ExpatParser instance at 0x0436D878>

Stack frame at level 7
======================
  File "C:\Program Files (x86)\wxGlade\xml_parse.py", line 94
  Function "parse_string(self=<xml_parse.CodeWriter instance at 0x0436D760>, source=<StringIO.StringIO instance at 0x043828A0>)"
  Source code context:
      
          def parse_string(self, source):
              source = StringIO.StringIO(source)
  ->          self.parser.parse(source)
              source.close()
      
          def setDocumentLocator(self, locator):
  Local variables:
  -> source (<type 'instance'>): <StringIO.StringIO instance at 0x043828A0>
  -> self (<type 'instance'>): <xml_parse.CodeWriter instance at 0x0436D760>

Stack frame at level 8
======================
  File "C:\Program Files (x86)\wxGlade\xml_parse.py", line 762
  Function "__init__(self=<xml_parse.CodeWriter instance at 0x0436D760>, writer=<perl_codegen.PerlCodeWriter object at 0x04098EB0>, input='<?xml version="1.0"?>\n<!-- generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016 -->\n\n<application class="" encoding="CP1252" for_version="3.0" header_extension=".h" indent_amount="4" indent_symbol="space" is_template="0" language="perl" name="" option="0" overwrite="1" path="wxglade_out.pl" source_extension=".cpp" top_window="" use_gettext="1" use_new_namespace="1">\n</application>\n', from_string=True, out_path=None, preview=False, class_names=set([]))"
  Source code context:
                  self.class_names = set()
      
              if from_string:
  ->              self.parse_string(input)
              else:
                  inputfile = None
                  try:
  Local variables:
  -> preview (<type 'bool'>): False
  -> out_path (<type 'NoneType'>): None
  -> self (<type 'instance'>): <xml_parse.CodeWriter instance at 0x0436D760>
  -> writer (<class 'perl_codegen.PerlCodeWriter'>): <perl_codegen.PerlCodeWriter object at 0x04098EB0>
  -> input (<type 'str'>): '<?xml version="1.0"?>\n<!-- generated by wxGlade 0.7.2 on Mon Jun 13 16:39:00 2016 -->\n\n<application class="" encoding="CP1252" for_version="3.0" header_extension=".h" indent_amount="4" indent_symbol="space" is_template="0" language="perl" name="" option="0" overwrite="1" path="wxglade_out.pl" source_extension=".cpp" top_window="" use_gettext="1" use_new_namespace="1">\n</application>\n'
  -> from_string (<type 'bool'>): True
  -> class_names (<type 'set'>): set([])

Stack frame at level 9
======================
  File "C:\Program Files (x86)\wxGlade\application.py", line 734
  Function "generate_code(self=<application.Application object at 0x0434DB70>, *args=(<wx._core.CommandEvent; proxy of <Swig Object of type 'wxCommandEvent *' at 0x5ceae4> >,), **kwds={})"
  Source code context:
      
                  CodeWriter(common.code_writers[codewriter], tmp_xml.getvalue(),
                             True, preview=preview, out_path=out_path,
  ->                         class_names=class_names)
      
                  # restore saved settings
                  if preview and codewriter == 'python':
  Local variables:
  -> tmp_xml (<class 'misc.UnicodeStringIO'>): <misc.UnicodeStringIO object at 0x037CF090>
  -> kwds (<type 'dict'>): {}
  -> out_path (<type 'NoneType'>): None
  -> self (<class 'application.Application'>): <application.Application object at 0x0434DB70>
  -> args (<type 'tuple'>): (<wx._core.CommandEvent; proxy of <Swig Object of type 'wxCommandEvent *' at 0x5ceae4> >,)
  -> CodeWriter (<type 'classobj'>): <class xml_parse.CodeWriter at 0x03541B90>
  -> codewriter (<type 'str'>): 'perl'
  -> inst (<class 'xml_parse.XmlParsingError'>): XmlParsingError(u"[Errno 13] Permission denied: u'wxglade_out.pl' (line: 5, column: 0)",)
  -> preview (<type 'bool'>): False
  -> class_names (<type 'set'>): set([])
~~~

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-06-14 16:59 UTC**
- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -1,5 +1,6 @@
 While trying to generate Perl or Python or C++ code, I get this message on 2 different PC:
 
+~~~
 ERROR   : An internal error occurred while Generate Code
 An unexpected error occurred!
 
@@ -201,3 +202,4 @@
   -> inst (<class 'xml_parse.XmlParsingError'>): XmlParsingError(u"[Errno 13] Permission denied: u'wxglade_out.pl' (line: 5, column: 0)",)
   -> preview (<type 'bool'>): False
   -> class_names (<type 'set'>): set([])
+~~~

~~~~

- **assigned_to**: Carsten Grohmann
- **Affected Version**: wxGlade version: 0.7.2 --> 0.7.X on Windows

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-06-14 16:59 UTC**
Hi Anciaux,

the error message is a misleading. The primary error is a permission error during saving the file "wxglade_out.pl".

Please select a writable directory to store the file into.

I'll improve the error message to be more precise.

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Commented by frank6363 on 2016-06-14 21:01 UTC**
Thanks a lot, it works fine and I can use this Wonderful tool !

**Comment by DietmarSchwertberger** [comment]
**Updated by cgrohmann on 2016-06-24 06:25 UTC**
- **status**: open --> closed

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-06-24 06:25 UTC**
Hi Frank,

the issue as shown in this bug report is, that IOErrors like "Permission denied" are wrapped in a XMLProcessingError. That's the bug, because IOErrors will be reported to the user already and the user can take corrective actions. The XMLProcessingErrors are internal errors and shouldn't occur.

The bug is fixed in revision 1957:cb3681b118dd "Don't wrap up IOError in XmlParsingError (sf bug #189)" in the current development branch. The fix will be part of the next official release.

You can download and use the fixed version from the current development branch directly: https://bitbucket.org/wxglade/wxglade/get/default.tar.gz .

The installation from such a downloaded source is described in section "Installation" of the wxGlade manual at http://wxglade.sourceforge.net/manual/ch02s02.html .

Thank you for reporting this bug.
Carsten
